Allows players to name seed packs to keep track of modified plants

theo-3 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Adds a 'variant' var to seed packs which can be set by players with a pen or with the dna modifier. This variable is shown in the pack's name, in the seed extractor, in the dna modifier and in hydroponic trays' examine. Extracting seed packs from named seeds will keep the name.

## Why It's Good For The Game
Helps keep track of what modified plants do, for example a tomato plant labeled 'healing' and another labeled 'acid', or a gaia plant labeled 'perrenial densified chemicals'.

## Images of changes
![2021-07-17 14-23-00](https://user-images.githubusercontent.com/85434590/126037577-4a3931eb-f447-4d90-acad-859e9bfa99e8.gif)
The gif doesn't show the dna modifier UI because I messed up the window capture. Sorry.
![Capture1](https://user-images.githubusercontent.com/85434590/126038014-5584c221-0854-4d53-a66b-bd209a60b48e.PNG)
![Capture2](https://user-images.githubusercontent.com/85434590/126038015-78492638-40e3-4a38-af1d-2b6d13011f52.PNG)
![Capture1](https://user-images.githubusercontent.com/85434590/126043095-5f72f75c-fc54-40be-9076-b359473fc479.PNG)




## Changelog
:cl:
add: Adds a 'variant' var to seed packets which can be set by using a pen on seeds or a tray, or from the dna modifier. It shows up in the packet's name, in plant trays examine, in the seed extractor, in the dna modifier and in harvested plants. This is intended to help keep track of what modified plants do, such as wheat labeled 'healing'.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
<!-- If a PR has no impact on players (i.e. a code refactor that does not change functionality) then the entire Changelog heading and contents can be removed. -->

dearmochi commented on code/modules/hydroponics/gene_modder.dm:
```diff
@@ -391,6 +398,19 @@
+	else if(href_list["set_v"])
+		if(!seed)
+			return
+		var/V = sanitize(stripped_input(usr, "Choose variant name:", "Plant Variant Naming", seed.variant, MAX_MESSAGE_LEN))
```
Might want to use a max length lower than 1024, or you could end up with absurdly long names that fill up the chat. Perhaps 64 characters is saner.

theo-3 replied:
Thanks, I copy pasted that from another file and didn't know what the max len did

dearmochi commented on code/modules/hydroponics/seeds.dm:
```diff
@@ -317,16 +320,32 @@
 
 		return
+	if(istype(O, /obj/item/pen))
+		var/V = sanitize(stripped_input(user, "Choose variant name:", "Plant Variant Naming", variant, MAX_MESSAGE_LEN))
```
Ditto

dearmochi commented on code/modules/hydroponics/seeds.dm:
```diff
@@ -317,16 +320,32 @@
+		else if(V != "")
+			user.show_message("<span class='info'>You designate the [plantname] as the [V] variant.</span>")
+		variant = V
+		apply_variant_name()
```
```suggestion
		apply_variant_name()
		return
```

So it doesn't proceed to the default behaviour

dearmochi commented on code/modules/hydroponics/seeds.dm:
```diff
@@ -317,16 +320,32 @@
+	var/P = findtext(name, "\[")
+	var/V = (variant == "") ? "" : (" \[" + variant + "]") // If we have a non-empty variant add it to the name
+	if (P != 0)
+		name = copytext(name, 1, P - 1) + V // If there was already a variant in the name remove it
```
Consider the `initial(varname)` proc to get the initial name of a datum, would be better than text manipulation

theo-3 replied:
Wouldn't that delete the 'experimental' from modified seeds?

dearmochi replied:
Good point, this is fine then

theo-3 commented:
Thanks for the help!

theo-3 commented:
I improved the apply name proc. It uses initial(name) and checks if 'experimental' was there, and also reapplies labels. This also means it won't get confused if you had a label like '([][])' present before setting a variant, which would have messed up the name before.

Wojz commented:
Good idea.

cluwneythescourge commented:
man, I've been waiting for something like this forever.  Does the modified name show on the actual plant?   Like, if you grow a "healing" tomato seed, are the plants labeled?

theo-3 commented:
The actual tomatoes are not labeled, since I thought it'd be weird to see a label on plants like that, but I could make it show up there too

theo-3 commented:
There, I added the label to growns too

theo-3 commented:
They're separate in plant bags now too, just tested it

theo-3 commented:
How can I tell whether a player pressed cancel or entered an empty string? Right now pressing cancel deletes the variant instead of doing nothing

dearmochi commented:
> How can I tell whether a player pressed cancel or entered an empty string? Right now pressing cancel deletes the variant instead of doing nothing

Check if the result of the input is `null` or otherwise empty.

theo-3 commented:
Yeah, I tried using !V, but that prevents a player from deleting a variant by entering an empty string. How do I tell the difference between canceling and typing an empty string?

dearmochi commented:
Cancelling specifically returns `null` while an empty string is, well, an empty string.

```dm
if(isnull(V))
	// Cancelled
else if(!length(V))
	// Empty string
```

theo-3 commented:
Huh. There's a difference between isnull(V) and V == null?

theo-3 commented:
`var/V = stripped_input(user, "Choose variant name:", "Plant Variant Naming", variant, 64)
		if((!Adjacent(user)) || isnull(V))
			[return`](url)
This isn't returning, pressing cancel deletes the variant

theo-3 commented:
![Capture](https://user-images.githubusercontent.com/85434590/126947186-08368253-92a5-40d2-8194-a7946b4306fe.PNG)
This isn't working. Pressing cancel still deletes the variant

theo-3 commented:
Yep, it works now. The weird thing is I look at the pandemic machine and it doesn't do this. It just checks for null text without delaying sanitize.

theo-3 commented:
Nevermind, the pandemic machine just makes unlabeled bottles if you cancel

theo-3 commented:
There, now cancel works right

theo-3 commented:
I moved the variant name prompt to a proc to remove code reuse from the gene modder and added naming to trays too. I'm using an adj parameter to get the proc to call Adjacent on the correct object. Is that the right way to do that?

Kyep reviewed (approved):
While I am fine with the concept, I wonder if this requires TGUI. Including raw HTML seems... not great.

theo-3 commented:
> While I am fine with the concept, I wonder if this requires TGUI. Including raw HTML seems... not great.

All botany machines use raw HTML, don't they?

AffectedArc07 commented on code/modules/hydroponics/grown.dm:
```diff
@@ -42,6 +42,8 @@
 		seed.prepare_result(src)
 		transform *= TRANSFORM_USING_VARIABLE(seed.potency, 100) + 0.5 //Makes the resulting produce's sprite larger or smaller based on potency!
 		add_juice()
+		if(seed.variant && seed.variant != "")
```
A string of `""` evalutes as false.
![image](https://user-images.githubusercontent.com/25063394/129193224-eab17828-7f65-4990-bb5f-ecd8af6e63ae.png)
The second check here (and all other places) is redundant.

theo-3 replied:
Done

AffectedArc07 commented:
> While I am fine with the concept, I wonder if this requires TGUI. Including raw HTML seems... not great.

I am not going to enforce a TGUI conversion in this PR since it is outside the scope.

theo-3 commented on code/modules/hydroponics/gene_modder.dm:
```diff
@@ -391,6 +398,15 @@
+	else if(href_list["del_v"])
+		if(!seed)
+			return
+		seed.variant = ""
```
Alright, I replaced all empty strings with null. Also fixed a bunch of == "" that wouldn't find null

theo-3 commented:
Uh. How did this break the tests?

theo-3 commented:
Is this because I force-changed the commit message?

theo-3 commented:
Let's try this again

theo-3 commented:
Alright, I tested it locally and it seems to work. @AffectedArc07 Is it good now?

theo-3 commented:
Quick question: is the stripped_input function supposed to convert nulls to empty strings? It makes it impossible to tell the difference between canceling and empty strings. I tested this with the pandemic machine, which seems to check for cancel too (`if(name == null || wait)`), and canceling creates a bottle named 'culture bottle' instead of doing nothing.
